animated-transformer/src/lib/transformer/gpt2.ts

```diff
@@ -0,0 +1,567 @@
+    .applyPointWiseTfFn(gelu)
+    .contract(ff2.w, ['hiddenRep'])
+    .pointwiseAdd(ff2.b);
+
+  // Dropout before residual connection.
+  seqOutput = dropout(
+    spec.dropoutRate,
+    seqOutput,
+    generator.random()
+  );
+
+  // Residual after MLP.
+  if (spec.residuals) {
+    seqOutput = seqOutput.pointwiseAdd(
+      headsReductionAfterResidual.rename('inputRepToFF', 'inputRep')
+    );
+  }
+
+  return {
+    seqInput,
+    keys,
+    queries,
+    attention,
+    values,
+    attendedValues,
+    // Note: this inputToFF is after the layer norm.
+    inputToFF: inputToFF.rename('inputRepToFF', 'inputRep'),
+    seqOutput,
+  };
+}
+
+export function initDecoderParams(config: Config): TransformerParams {
+  const { spec, init } = config;
+  // const paramInitializerConfig = config.init;
+  const layers: AttnHeadParams[] = spec.layers.map((layerSpec) => {
+    const attnHeadSpec: AttnHeadParamSpec = {
+      inputRep: spec.inputRep,
+      kq: spec.kqvRep,
+      heads: layerSpec.nHeads,
+      value: spec.kqvRep,
+      layerNormPreAttention: layerSpec.layerNormPreAttention,
+      layerNormHeadsProjection: layerSpec.layerNormHeadsProjection,
+      // addLayerNormBias: AttentionIsAllYouNeed = true; T5 = false.
+      addLayerNormBias: layerSpec.addLayerNormBias,
+    };
+    return initAttnHeadParams(attnHeadSpec, init);
+  });
+  const tokenEmbedding = makeTruncNormal(
+    {
+      tokenId: config.tokenRep.tokens.length,
+      inputRep: spec.inputRep,
+    },
+    init
+  );
+
+  let transformerParams: TransformerParams = {
+    tokenEmbedding: tokenEmbedding,
+    layers: layers
+  };
+
+  if (spec.addPosEmbeddings) {
+    const posEmbedding = makeTruncNormal({
+      posId: spec.posEncodingSeqLength,
+      inputRep: spec.inputRep,
+    });
+    transformerParams.posEmbedding = posEmbedding;
+  }
+
+  if (spec.layerNorm) {
+    const layerNormParams = initLayerNormParamsWithDims(
+      spec.addLayerNormBias, { 'inputRep': spec.inputRep });
+    transformerParams.layerNorm = layerNormParams
+  }
+
+  return transformerParams;
+}
+
+export type TransformerComputation = {
+  layers: BatchAttnHeadCompututation[];
+};
+
+export function computeTransformer(
+  model: {
+    config: { spec: TransformerParamSpec };
+    params: TransformerParams;
+  },
+  seqInput: GTensor<'batch' | 'pos' | 'inputRep'>,
+  generator: RandomStream
+): TransformerComputation {
+  const compute: TransformerComputation = { layers: [] };
+  let currentLayerInput = dropout(model.config.spec.computeSpec.dropoutRate,
+    seqInput, generator.random());
+  model.params.layers.forEach((layerParams, i) => {
+    const layerCompute = computeAttnHead(
+      model.config.spec.layers[i].computeSpec,
+      layerParams,
+      currentLayerInput,
+      generator
+    );
+    compute.layers.push(layerCompute);
+    currentLayerInput = layerCompute.seqOutput;
+  });
+
+  // TODO(@aliciafmachado): Hacky way to apply layer norm after the final layer.
```

Comment:
Any thoughts on this? Not sure if there would be a easy way to hook this in, and we have been only using the last layer in compute for now anyway.

Reply:
I'm not sure; but I think it's ok, definitely ok for now.